File: packages/spotify_sdk_android/android/src/main/kotlin/de/minimalme/spotify_sdk/SpotifySdkPlugin.kt
package de.minimalme.spotify_sdk

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry
import de.minimalme.spotify_sdk.handlers.AuthHandler
import de.minimalme.spotify_sdk.handlers.ImageHandler
import de.minimalme.spotify_sdk.handlers.LibraryHandler
import de.minimalme.spotify_sdk.handlers.PlayerHandler
import de.minimalme.spotify_sdk.subscriptions.ConnectionStatusChannel

class SpotifySdkPlugin : MethodCallHandler, FlutterPlugin, ActivityAware, PluginRegistry.ActivityResultListener {

    private lateinit var methodChannel: MethodChannel
    val remoteManager = RemoteManager()
    val authHandler = AuthHandler(remoteManager)
    val playerHandler = PlayerHandler(remoteManager)
    val libraryHandler = LibraryHandler(remoteManager)
    val imageHandler = ImageHandler(remoteManager)

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        remoteManager.applicationContext = binding.applicationContext

        methodChannel = MethodChannel(binding.binaryMessenger, SpotifySdkConstants.CHANNEL_SPOTIFY_SDK)
        methodChannel.setMethodCallHandler(this)

        remoteManager.playerContextChannel = EventChannel(binding.binaryMessenger, SpotifySdkConstants.CHANNEL_PLAYER_CONTEXT)
        remoteManager.playerStateChannel = EventChannel(binding.binaryMessenger, SpotifySdkConstants.CHANNEL_PLAYER_STATE)
        remoteManager.capabilitiesChannel = EventChannel(binding.binaryMessenger, SpotifySdkConstants.CHANNEL_CAPABILITIES)
        remoteManager.userStatusChannel = EventChannel(binding.binaryMessenger, SpotifySdkConstants.CHANNEL_USER_STATUS)
        remoteManager.connectionStatusChannel = EventChannel(binding.binaryMessenger, SpotifySdkConstants.CHANNEL_CONNECTION_STATUS)

        remoteManager.connectionStatusChannel?.setStreamHandler(ConnectionStatusChannel(remoteManager.connStatusEventChannel))
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        remoteManager.applicationContext = null
        methodChannel.setMethodCallHandler(null)

        remoteManager.playerContextChannel?.setStreamHandler(null)
        remoteManager.playerStateChannel?.setStreamHandler(null)
        remoteManager.capabilitiesChannel?.setStreamHandler(null)
        remoteManager.userStatusChannel?.setStreamHandler(null)
        remoteManager.connectionStatusChannel?.setStreamHandler(null)

        remoteManager.playerContextChannel = null
        remoteManager.playerStateChannel = null
        remoteManager.capabilitiesChannel = null
        remoteManager.userStatusChannel = null
        remoteManager.connectionStatusChannel = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        binding.addActivityResultListener(this)
        remoteManager.applicationActivity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        remoteManager.applicationActivity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        remoteManager.applicationActivity = binding.activity
    }

    override fun onDetachedFromActivity() {
        remoteManager.applicationActivity = null
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            // Auth & Session
            SpotifySdkConstants.METHOD_CONNECT_TO_SPOTIFY -> authHandler.connectToSpotify(
                call.argument(SpotifySdkConstants.PARAM_CLIENT_ID),
                call.argument(SpotifySdkConstants.PARAM_REDIRECT_URL),
                result
            )
            SpotifySdkConstants.METHOD_GET_ACCESS_TOKEN -> authHandler.getAccessToken(
                call.argument(SpotifySdkConstants.PARAM_CLIENT_ID),
                call.argument(SpotifySdkConstants.PARAM_REDIRECT_URL),
                call.argument(SpotifySdkConstants.PARAM_SCOPE),
                result
            )
            SpotifySdkConstants.METHOD_GET_SWAP_TOKEN -> authHandler.getSwapToken(
                call.argument(SpotifySdkConstants.PARAM_CLIENT_ID),
                call.argument(SpotifySdkConstants.PARAM_REDIRECT_URL),
                call.argument(SpotifySdkConstants.PARAM_SCOPE),
                result
            )
            SpotifySdkConstants.METHOD_IS_SPOTIFY_INSTALLED -> authHandler.isSpotifyInstalled(result)
            SpotifySdkConstants.METHOD_DISCONNECT_FROM_SPOTIFY -> authHandler.disconnectFromSpotify(result)

            // Playback & Controls
            SpotifySdkConstants.METHOD_SWITCH_TO_LOCAL_DEVICE -> playerHandler.switchToLocalDevice(result)
            SpotifySdkConstants.METHOD_GET_CROSSFADE_STATE -> playerHandler.getCrossfadeState(result)
            SpotifySdkConstants.METHOD_GET_PLAYER_STATE -> playerHandler.getPlayerState(result)
            SpotifySdkConstants.METHOD_PLAY -> playerHandler.play(call.argument(SpotifySdkConstants.PARAM_SPOTIFY_URI), result)
            SpotifySdkConstants.METHOD_PAUSE -> playerHandler.pause(result)
            SpotifySdkConstants.METHOD_QUEUE_TRACK -> playerHandler.queue(call.argument(SpotifySdkConstants.PARAM_SPOTIFY_URI), result)
            SpotifySdkConstants.METHOD_RESUME -> playerHandler.resume(result)
            SpotifySdkConstants.METHOD_SEEK_TO -> playerHandler.seekTo((call.argument<Number>(SpotifySdkConstants.PARAM_POSITIONED_MILLISECONDS))?.toLong(), result)
            SpotifySdkConstants.METHOD_SEEK_TO_RELATIVE_POSITION -> playerHandler.seekToRelativePosition((call.argument<Number>(SpotifySdkConstants.PARAM_RELATIVE_MILLISECONDS))?.toLong(), result)
            SpotifySdkConstants.METHOD_SET_PODCAST_PLAYBACK_SPEED -> playerHandler.setPodcastPlaybackSpeed((call.argument<Number>(SpotifySdkConstants.PARAM_PODCAST_PLAYBACK_SPEED))?.toDouble(), result)
            SpotifySdkConstants.METHOD_SKIP_NEXT -> playerHandler.skipNext(result)
            SpotifySdkConstants.METHOD_SKIP_PREVIOUS -> playerHandler.skipPrevious(result)
            SpotifySdkConstants.METHOD_SKIP_TO_INDEX -> playerHandler.skipToIndex(call.argument(SpotifySdkConstants.PARAM_SPOTIFY_URI), call.argument(SpotifySdkConstants.PARAM_TRACK_INDEX), result)
            SpotifySdkConstants.METHOD_TOGGLE_SHUFFLE -> playerHandler.toggleShuffle(result)
            SpotifySdkConstants.METHOD_SET_SHUFFLE -> playerHandler.setShuffle(call.argument(SpotifySdkConstants.PARAM_SHUFFLE), result)
            SpotifySdkConstants.METHOD_TOGGLE_REPEAT -> playerHandler.toggleRepeat(result)
            SpotifySdkConstants.METHOD_SET_REPEAT_MODE -> playerHandler.setRepeatMode(call.argument(SpotifySdkConstants.PARAM_REPEAT_MODE), result)

            // User Library & Capabilities
            SpotifySdkConstants.METHOD_ADD_TO_LIBRARY -> libraryHandler.addToUserLibrary(call.argument(SpotifySdkConstants.PARAM_SPOTIFY_URI), result)
            SpotifySdkConstants.METHOD_REMOVE_FROM_LIBRARY -> libraryHandler.removeFromUserLibrary(call.argument(SpotifySdkConstants.PARAM_SPOTIFY_URI), result)
            SpotifySdkConstants.METHOD_GET_CAPABILITIES -> libraryHandler.getCapabilities(result)
            SpotifySdkConstants.METHOD_GET_LIBRARY_STATE -> libraryHandler.getLibraryState(call.argument(SpotifySdkConstants.PARAM_SPOTIFY_URI), result)

            // Cover Art Images
            SpotifySdkConstants.METHOD_GET_IMAGE -> imageHandler.getImage(call.argument(SpotifySdkConstants.PARAM_IMAGE_URI), call.argument(SpotifySdkConstants.PARAM_IMAGE_DIMENSION), result)

            else -> result.notImplemented()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: android.content.Intent?): Boolean {
        return authHandler.onActivityResult(requestCode, resultCode, data)
    }
}

File: packages/spotify_sdk_ios/ios/spotify_sdk_ios/Classes/SpotifySdkPlugin.swift
import Flutter
import SpotifyiOS

public class SpotifySdkPlugin: NSObject, FlutterPlugin {
    private static var instance = SpotifySdkPlugin()

    private let remoteManager = RemoteManager.shared
    private lazy var authHandler = AuthHandler(remoteManager: remoteManager)
    private lazy var playerHandler = PlayerHandler(remoteManager: remoteManager)
    private lazy var libraryHandler = LibraryHandler(remoteManager: remoteManager)
    private lazy var imageHandler = ImageHandler(remoteManager: remoteManager)

    public static func register(with registrar: FlutterPluginRegistrar) {
        guard RemoteManager.playerStateChannel == nil else {
            return
        }
        let spotifySDKChannel = FlutterMethodChannel(name: SpotifySdkConstants.channelSpotifySdk, binaryMessenger: registrar.messenger())
        let connectionStatusChannel = FlutterEventChannel(name: SpotifySdkConstants.channelConnectionStatus, binaryMessenger: registrar.messenger())
        RemoteManager.connectionStatusChannel = connectionStatusChannel
        RemoteManager.playerStateChannel = FlutterEventChannel(name: SpotifySdkConstants.channelPlayerState, binaryMessenger: registrar.messenger())
        RemoteManager.playerContextChannel = FlutterEventChannel(name: SpotifySdkConstants.channelPlayerContext, binaryMessenger: registrar.messenger())
        RemoteManager.capabilitiesChannel = FlutterEventChannel(name: SpotifySdkConstants.channelCapabilities, binaryMessenger: registrar.messenger())
        RemoteManager.userStatusChannel = FlutterEventChannel(name: SpotifySdkConstants.channelUserStatus, binaryMessenger: registrar.messenger())

        registrar.addApplicationDelegate(instance)

        if #available(iOS 13.0, *) {
            let selector = NSSelectorFromString("addSceneDelegate:")
            if (registrar as AnyObject).responds(to: selector) {
                (registrar as AnyObject).perform(selector, with: instance)
            }
        }
        registrar.addMethodCallDelegate(instance, channel: spotifySDKChannel)

        instance.remoteManager.connectionStatusHandler = ConnectionStatusHandler()
        connectionStatusChannel.setStreamHandler(instance.remoteManager.connectionStatusHandler)

        instance.remoteManager.capabilitiesHandler = CapabilitiesHandler()
        RemoteManager.capabilitiesChannel?.setStreamHandler(instance.remoteManager.capabilitiesHandler)

        instance.remoteManager.userStatusHandler = UserStatusHandler()
        RemoteManager.userStatusChannel?.setStreamHandler(instance.remoteManager.userStatusHandler)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        // Auth & Session
        case SpotifySdkConstants.methodConnectToSpotify:
            authHandler.connectToSpotify(call: call, result: result)
        case SpotifySdkConstants.methodGetAccessToken, SpotifySdkConstants.methodGetSwapToken:
            authHandler.getAccessTokenOrSwapToken(call: call, result: result)
        case SpotifySdkConstants.methodIsSpotifyInstalled:
            authHandler.isSpotifyInstalled(result: result)
        case SpotifySdkConstants.methodDisconnectFromSpotify:
            authHandler.disconnect(result: result)

        // Playback & Controls
        case SpotifySdkConstants.methodGetPlayerState:
            playerHandler.getPlayerState(result: result)
        case SpotifySdkConstants.methodPlay:
            playerHandler.play(call: call, result: result)
        case SpotifySdkConstants.methodPause:
            playerHandler.pause(result: result)
        case SpotifySdkConstants.methodResume:
            playerHandler.resume(result: result)
        case SpotifySdkConstants.methodQueueTrack:
            playerHandler.queueTrack(call: call, result: result)
        case SpotifySdkConstants.methodSkipNext:
            playerHandler.skipNext(result: result)
        case SpotifySdkConstants.methodSkipPrevious:
            playerHandler.skipPrevious(result: result)
        case SpotifySdkConstants.methodSkipToIndex:
            playerHandler.skipToIndex(call: call, result: result)
        case SpotifySdkConstants.methodSeekTo:
            playerHandler.seekTo(call: call, result: result)
        case SpotifySdkConstants.methodSeekToRelativePosition:
            playerHandler.seekToRelativePosition(call: call, result: result)
        case SpotifySdkConstants.methodGetCrossfadeState:
            playerHandler.getCrossfadeState(result: result)
        case SpotifySdkConstants.methodSetShuffle:
            playerHandler.setShuffle(call: call, result: result)
        case SpotifySdkConstants.methodToggleShuffle:
            playerHandler.toggleShuffle(result: result)
        case SpotifySdkConstants.methodSetRepeatMode:
            playerHandler.setRepeatMode(call: call, result: result)
        case SpotifySdkConstants.methodToggleRepeat:
            playerHandler.toggleRepeat(result: result)
        case SpotifySdkConstants.methodSwitchToLocalDevice:
            playerHandler.switchToLocalDevice(result: result)

        // User Library & Capabilities
        case SpotifySdkConstants.methodAddToLibrary:
            libraryHandler.addToLibrary(call: call, result: result)
        case SpotifySdkConstants.methodRemoveFromLibrary:
            libraryHandler.removeFromLibrary(call: call, result: result)
        case SpotifySdkConstants.methodGetCapabilities:
            libraryHandler.getCapabilities(result: result)
        case SpotifySdkConstants.methodGetLibraryState, SpotifySdkConstants.getLibraryState:
            libraryHandler.getLibraryState(call: call, result: result)

        // Cover Art Images
        case SpotifySdkConstants.methodGetImage:
            imageHandler.getImage(call: call, result: result)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    public func application(_ application: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey: Any] = [:]) -> Bool {
        setAccessTokenFromURL(url: url)
        return true
    }

    public func application(_ application: UIApplication, continue userActivity: NSUserActivity, restorationHandler: @escaping ([Any]) -> Void) -> Bool {
        guard userActivity.activityType == NSUserActivityTypeBrowsingWeb,
              let url = userActivity.webpageURL else {
            return false
        }
        setAccessTokenFromURL(url: url)
        return false
    }

    private func setAccessTokenFromURL(url: URL) {
        guard let appRemote = remoteManager.appRemote else {
            remoteManager.connectionStatusHandler?.connectionResult?(FlutterError(code: "errorConnection", message: "AppRemote is null", details: nil))
            remoteManager.connectionStatusHandler?.tokenResult?(FlutterError(code: "errorConnection", message: "AppRemote is null", details: nil))
            remoteManager.connectionStatusHandler?.connectionResult = nil
            remoteManager.connectionStatusHandler?.tokenResult = nil
            return
        }

        guard let token = appRemote.authorizationParameters(from: url)?[SPTAppRemoteAccessTokenKey] else {
            remoteManager.connectionStatusHandler?.connectionResult?(FlutterError(code: "authenticationTokenError", message: appRemote.authorizationParameters(from: url)?[SPTAppRemoteErrorDescriptionKey], details: nil))
            remoteManager.connectionStatusHandler?.tokenResult?(FlutterError(code: "authenticationTokenError", message: appRemote.authorizationParameters(from: url)?[SPTAppRemoteErrorDescriptionKey], details: nil))
            remoteManager.connectionStatusHandler?.connectionResult = nil
            remoteManager.connectionStatusHandler?.tokenResult = nil
            return
        }

        appRemote.connectionParameters.accessToken = token
        appRemote.connect()
    }
}

@available(iOS 13.0, *)
extension SpotifySdkPlugin: UIWindowSceneDelegate {
    public func scene(_ scene: UIScene, openURLContexts URLContexts: Set<UIOpenURLContext>) {
        guard let url = URLContexts.first?.url else { return }
        setAccessTokenFromURL(url: url)
    }

    public func scene(_ scene: UIScene, continue userActivity: NSUserActivity) {
        guard userActivity.activityType == NSUserActivityTypeBrowsingWeb,
              let url = userActivity.webpageURL else {
            return
        }
        setAccessTokenFromURL(url: url)
    }
}
